Feature/api

janbritz opened the pull request:
Funktionsfähiger Entwurf der API (mit Ausnahme von zwei Routes).

Über die Routes kann mit dem Decorator `ensure_package_and_question_state_exists` auf den Pfad des Pakets und des `QuestionState`s zugegriffen werden.

MartinGauk commented on questionpy_server/web.py:
```diff
@@ -0,0 +1,192 @@
+                    # set cookies with hashes
+                    response.set_cookie("package_hash", package_hash, max_age=0)
+                    response.set_cookie("question_state_hash", question_state_hash, max_age=0)
+                    response.set_cookie("request_id", request_id, max_age=0)
```
Die Cookies verstehe ich nicht. Die Informationen sind ja alle im JSON enthalten, die in jedem Request übermittelt werden. Es wird daher auch kein request_cache benötigt.

MartinGauk commented on questionpy_server/settings.py:
```diff
@@ -30,10 +35,29 @@ def __call__(self, settings: BaseSettings) -> Dict[str, Any]:
+
+
+class CollectorSettings(BaseModel):
+    local_directory: str = 'questionpy_server/packages'
```
Alle drei Ordner sollten voreingestellt woanders liegen. Mitten im Code erscheint mir unpassend. Der Cache könnte auf oberster Ebene stehen. Für `local_directory` sollte es vermutlich besser gar keinen default geben.

MartinGauk commented on questionpy_server/cache.py:
```diff
@@ -0,0 +1,160 @@
+
+        # Get files and sort them by access time.
+        file_stats = self.files()
+        file_stats_sorted = sorted(file_stats.items(), key=lambda item: item[1].st_atime_ns)
```
Das ist zu sehr vom Dateisystem abhängig (oft wird mit `noatime` gemountet). Ich würde die Cache-Keys (also die Dateinamen) bei Programmstart in einen `OrderedDict` einlesen und aktuell halten. Beim Zugriff auf einen Inhalt wird `move_to_end` aufgerufen.
Wenn direkt nach dem Programmstart Elemente aus dem Cache entfernt werden müssen, arbeitet der Cache zwar zunächst nicht nach LRU, aber das ist vertretbar.

MHajoha commented on questionpy_server/api/models.py:
```diff
@@ -0,0 +1,353 @@
+
+
+class QuestionCreateArguments(BaseModel):
+    class Config:
+        extra = Extra.forbid
```
Ich habe ja selbst in der OpenAPI spec überall `additionalProperties: false` geschrieben, und das ist so auch korrekt abgebildet. Bei näherem Darübernachdenken halte ich das aber doch für keine gute Idee, da es das Erweitern von Modellen zu einer brechenden Änderung macht: Sendet der Client ein neues Feld, das der Server noch nicht kennt, failt der Request, und andersherum ebenso.

Das Standart-Verhalten von pydantic, unbekanntes einfach wegzuwefen, halte ich für besser.

As for pylint: Anscheinend kann man die ´too-few-public-methods´-Regel nicht für bestimmte Klassennamen ausschalten. Deshalb würde ich fast vorschlagen, sie global zu deaktivieren. Alternativ könnte `class Config(pydantic.BaseConfig)` als Workaround Abhilfe schaffen. Das sollte (nicht ausprobiert) gleichbedeutend sein, aber da `BaseConfig` zufällig 2 öffentliche Methoden hat, schlägt pylint nicht an.

Oder eben `pylint: disable=too-few-public-methods` :)

janbritz replied:
Definitiv, allein wegen der möglichen Kompatibilitätsprobleme, sollten wir zusätzliche Felder erlauben.

Ich werde jetzt erstmal die pylint-Regel hinzufügen!

MHajoha commented on questionpy_server/api/models.py:
```diff
@@ -0,0 +1,353 @@
+from pydantic import BaseModel, Extra, Field, FilePath, HttpUrl, Json
+
+
+class PackageType(Enum):
+    library = 'library'
+    questiontype = 'questiontype'
+    question = 'question'
```
Wir sollten die Naming Convention der Enums aneinander angleichen. Ich habe eine schwache Vorliebe für lower_snake_case wie hier, aber UPPER_SNAKE_CASE wie unten wäre auch okay.

janbritz replied:
Mein Favorit ist auch snake_case - ich habe die Enums angepasst.

MHajoha commented on questionpy_server/api/models.py:
```diff
@@ -0,0 +1,353 @@
+
+class HiddenElement(BaseModel):
+    kind: ElementKind = Field(ElementKind.hidden, const=True)
+    name: FormName
+    value: str
+
+
+class GroupElement(BaseModel):
+    kind: ElementKind = Field(ElementKind.group, const=True)
+    label: FormLabel
+    elements: List[FormElement]
+
+
+class FormElement(BaseModel):
+    __root__: Union[
+        StaticTextElement,
+        TextInputElement,
+        CheckboxElement,
+        CheckboxGroupElement,
+        RadioGroupElement,
+        SelectElement,
+        HiddenElement,
+        GroupElement,
+    ]
+
+
+class Section(BaseModel):
+    class Config:
+        extra = Extra.forbid
+
+    header: str
+    elements: List[FormElement]
+
+
+class OptionsFormDefinition(BaseModel):
+    class Config:
+        extra = Extra.forbid
+
+    general: Optional[List[FormElement]]
+    sections: Optional[List[Section]]
```
Klau' dir das am besten einfach aus [`questionpy.form`](https://github.com/questionpy-org/questionpy-sdk/blob/dev/questionpy/form/_elements.py). Eigentlich können wir `questionpy` auch als Library einbinden, was hälst du davon? :thinking:

janbritz replied:
Das wäre sicher die schönste Art! Leider entstehen dadurch aktuell Probleme mit dem Package `pydantic-factories`, welches ich zur Erstellung von Fakedaten verwende. Das Package kann wohl aktuell nicht mit `Annotated` umgehen. Ich bin aber dabei eine Lösung zu erarbeiten.

MHajoha replied:
Vielleicht könnte man der Factory das beibringen, indem man get_provider_map overridet? https://github.com/Goldziher/pydantic-factories/blob/main/pydantic_factories/factory.py#L228

MHajoha commented on questionpy_server/cache.py:
```diff
@@ -0,0 +1,152 @@
+        self._files: OrderedDict[str, File] = OrderedDict()
+
+        self._directory: Path = Path(directory).resolve()
+        self._directory.mkdir(exist_ok=True)
```
Hier vielleicht auch die Elternverzeichnisse mit erstellen?

```suggestion
        self._directory.mkdir(exist_ok=True, parents=True)
```

janbritz replied:
Gerne können wir hier noch die Elternverzeichnisse erstellen - ich persönlich finde es beruhigender, wenn ein Fehler ausgegeben wird, falls diese nicht existieren.
So kann ich sicher sein, den richtigen Pfad in den Settings angegeben zu haben (vor allem bei relativen Pfaden).

MHajoha replied:
Alles klar :+1:

MHajoha commented on questionpy_server/cache.py:
```diff
@@ -0,0 +1,152 @@
+
+        return path
+
+    @property
+    def total_bytes(self) -> int:
+        return self._total_bytes
```
Am liebsten gar keine trivialen Getter und Setter, aber auf keinen Fall in Python :sweat_smile: 

Dann lieber das Attribut umbennenen

MartinGauk replied:
Das Attribut soll ja nur read-only von außerhalb sein. In dem Fall finde ich das schon in Ordnung.

MHajoha commented on questionpy_server/cache.py:
```diff
@@ -0,0 +1,152 @@
+
+        # Save the bytes on filesystem.
+        path = self._path(key)
+        assert size == path.open('wb').write(value)
```
1. Es fehlt hier ein `close()` call. Oder noch besser mit `with`.

   > Calling f.write() without using the with keyword or calling f.close() might result in the arguments of f.write() not being completely written to the disk, even if the program exits successfully.
   https://docs.python.org/3/tutorial/inputoutput.html#reading-and-writing-files
2. Nebeneffekte in `assert`-Statements halte ich nicht für super gut. Zwar werden sie in Python zwar standardmäßig ausgeführt, aber mit `-O` oder der Umgebungsvariable `PYTHONOPTIMIZE` werden sie ausgeschaltet.

```suggestion
        with path.open("wb") as file:
            written_bytes = file.write(value)
            assert size == written_bytes
```

janbritz replied:
1. Da hast du absolut Recht. Ich wollte ursprünglich die Funktion `path.write_bytes(value)` verwenden, welche die Datei automatisch öffnet und schließt - werde es anpassen.
2. Ich habe das `assert`-Statement mit einer Exception ersetzt.

MHajoha commented on questionpy_server/cache.py:
```diff
@@ -0,0 +1,152 @@
+
+    @property
+    def directory(self) -> Path:
+        # Ensures that the cache directory exists.
+        if not self._directory.is_dir():
+            self._directory.mkdir(parents=True, exist_ok=True)
+        return self._directory
```
Das Verzeichnis wird doch schon in `__init__` erstellt? Falls das Ziel ist, es nach potenzieller Löschung wieder zu erstellen bevor etwas hinzugefügt wird, würde ich das eher in `put` sehen und nicht in einer Property versteckt. Zumal diese Property eh nur innerhalb der Klasse benutzt wird.

janbritz replied:
Genau, nach einer Löschung sollte der Ordner wieder erstellt werden. Wie du schon erwähnt hast, wird die Funktion nur in `put` benötigt, weshalb das Auslagern der Logik in die Property nicht nötig ist - ich habe das angepasst.

MHajoha reviewed (changes requested):
Den Rest habe ich jetzt nicht geschafft, schaue ich mir dann Dienstag an.

MHajoha commented on questionpy_server/misc.py:
```diff
@@ -0,0 +1,24 @@
+def get_parameters_of_type(function: Callable, cls: type) -> List[Parameter]:
+    output = []
+    param: Parameter
+    for _, param in signature(function).parameters.items():
+        if isclass(param.annotation) and issubclass(param.annotation, cls):
+            output.append(param)
```
```suggestion
    for param in signature(function).parameters.values():
        if isclass(param.annotation) and issubclass(param.annotation, cls):
            output.append(param)
```

Wird der Typ von `param` ohne die annotation nicht korrekt inferred?

MHajoha commented on questionpy_server/settings.py:
```diff
@@ -21,7 +21,12 @@ def __call__(self, settings: BaseSettings) -> Dict[str, Any]:
+                'package_cache': dict(parser['cache-package']),
+                'question_state_cache': dict(parser['cache-question_state']),
+                'collector': dict(parser['collector'])
+            }
```
Können wir vielleicht die INI-Namen und die Feld-Namen aneinander angleichen? Dann könnte man das durch
```suggestion
            return {key: section for key, section in parser.items() if key != "DEFAULT"}
```

o.ä. ersetzen

MHajoha commented on questionpy_server/web.py:
```diff
@@ -0,0 +1,213 @@
+
+    main = cast(str, data.get('main'))
+    package = cast(Optional[FileField], data.get('package'))
+    question_state = cast(Optional[str], data.get('question_state'))
```
Sollten wir hier nicht checken, ob die Typen richtig sind, statt zu casten?

MHajoha commented on questionpy_server/web.py:
```diff
@@ -0,0 +1,213 @@
+                headers: Optional[LooseHeaders] = None) -> Optional[Path]:
+    """
+    Saves a package on `server` or retrieves it from cache if `package` is None, and returns `Path` to the package.
+    Raises `SizeError` if the `package` is too big for the cache.
```
Lügenpresse! `HTTPRequestEntityTooLarge` wird geworfen ;)

MHajoha commented on questionpy_server/web.py:
```diff
@@ -0,0 +1,213 @@
+
+
+def ensure_package_and_question_state_exists(_func: Optional[RouteHandler] = None, *,
+                                             headers: Optional[LooseHeaders] = None) \
```
Gibt es denn einen Use-Case dafür, hier Header für die Response mitgeben zu können?

MHajoha commented on tests/conftest.py:
```diff
@@ -1,18 +1,58 @@
+    QuestionStateCacheSettings
+
+
+class PackageNames(str, Enum):
+    complete = 'complete'
+    no_package = 'no_package'
+    no_question_state = 'no_question_state'
+    nothing = 'nothing'
```
```suggestion
class PackageName(str, Enum):
    complete = 'complete'
    no_package = 'no_package'
    no_question_state = 'no_question_state'
    nothing = 'nothing'
```

MHajoha commented on tests/test_cache.py:
```diff
@@ -0,0 +1,214 @@
+
+
+def test_remove(cache: FileLimitLRU, settings: Settings) -> None:
+    file, _ = settings.items.list[0]
+    cache.remove(file)
+    expected_total_bytes = settings.items.total_bytes - settings.items.bytes_per_item
+    assert not (cache.directory / file).is_file()
+    assert cache.total_bytes == expected_total_bytes
+
+    file, _ = settings.items.list[-1]
+    cache.remove(file)
+    expected_total_bytes = settings.items.total_bytes - 2 * settings.items.bytes_per_item
+    assert not (cache.directory / file).is_file()
+    assert cache.total_bytes == expected_total_bytes
+
+    with pytest.raises(FileNotFoundError):
+        cache.remove(file)
+    assert cache.total_bytes == expected_total_bytes
+
+    with pytest.raises(FileNotFoundError):
+        cache.remove('doesnotexist')
+    assert cache.total_bytes == expected_total_bytes
```
Mir wäre es ganz lieb, wenn wir uns auf einen test case pro Funktion beschränken. Die Funktion sollte dann einen aussagekräftigen Namen haben, z.B. `test_remove_should_raise_FileNotFoundError[_when_file_does_not_exist]`

MHajoha commented on tests/test_collector.py:
```diff
@@ -0,0 +1,100 @@
+        create_files(self.local_dir, self.local_files)
+
+
+@pytest.fixture
+def create_collector() -> Callable[[Settings], PackageCollector]:
+
+    def create_collector_function(settings: Settings) -> PackageCollector:
+        return PackageCollector(
+            local_dir=str(settings.local_dir),
+            cache=FileLimitLRU(
+                directory=str(settings.cache_dir),
+                max_bytes=10,
+                extension='.qpy',
+            ))
+
+    return create_collector_function
```
Kann das nicht eine einfache Funktion, keine Fixture, sein?

MartinGauk commented on .gitignore:
```diff
@@ -11,3 +11,6 @@ __pycache__
 /coverage.xml
+
+/questionpy_server/cache/*
+/questionpy_server/packages/*
```
Hier stehen noch die alten Ordner.

MartinGauk commented on config.example.ini:
```diff
@@ -10,3 +10,24 @@
+# The path of the cache (relative or absolute)
+#path = cache/packages
+
+[cache-question_state]
```
Ich würde bei den beiden sections Unterstriche nehmen. Dann lässt sich auch der Vorschlag von Max unten (in der settings.py) leicht umsetzen.

MartinGauk commented on questionpy_server/cache.py:
```diff
@@ -0,0 +1,146 @@
+            self._total_bytes += size
+
+            if self._total_bytes > self.max_bytes:
+                raise SizeError(f"Total bytes of {self.max_bytes} exceeded while initializing")
```
Das bedeutet ja, dass der Server nie wieder startet, wenn der Cache zu groß ist (weil man später mal die max. Cache-Größe verringert hat). Ich würde sagen, dass in so einem Fall die übrigen Dateien gelöscht werden.

janbritz committed:
feat: implementation of QPPE and routes

Routes return fake data; migration- and question-route are not
implemented yet.
